src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -350,4 +370,205 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
+        }
+    }
+
+    /// <summary>
+    /// Builds a global analyzer config from a series of configs
+    /// </summary>
+    internal class GlobalAnalyzerConfigBuilder
+    {
+        internal const string GlobalKey = "is_global";
+
+        PooledDictionary<string, ImmutableDictionary<string, (string, string)>.Builder> _values;
+        PooledDictionary<string, PooledDictionary<string, ArrayBuilder<string>>> _duplicates;
+
+        private GlobalAnalyzerConfigBuilder()
+        {
+            _values = PooledDictionary<string, ImmutableDictionary<string, (string, string)>.Builder>.GetInstance();
+            _duplicates = PooledDictionary<string, PooledDictionary<string, ArrayBuilder<string>>>.GetInstance();
+        }
+
+        /// <summary>
+        /// Filter out and merge any global configs
+        /// </summary>
+        /// <param name="analyzerConfigs">A list of <see cref="AnalyzerConfig"/> to filter out </param>
+        /// <param name="unsetAnalyzerKeys">Returns any global keys that were unset due to duplication</param>
+        /// <returns>An <see cref="ImmutableArray{AnalyzerConfig}"/> that contains the non-global configs and a single merged global config</returns>
+        internal static ImmutableArray<AnalyzerConfig> FilterGlobalConfigs<TList>(TList analyzerConfigs, out ImmutableArray<AnalyzerUnsetKey> unsetAnalyzerKeys) where TList : IReadOnlyCollection<AnalyzerConfig>
+        {
+            // don't do any actual work if there aren't any global configs
+            if (!containsGlobalConfigs())
+            {
+                unsetAnalyzerKeys = ImmutableArray<AnalyzerUnsetKey>.Empty;
+                return analyzerConfigs.ToImmutableArray();
+            }
+
+            ArrayBuilder<AnalyzerConfig> filteredConfigs = ArrayBuilder<AnalyzerConfig>.GetInstance(analyzerConfigs.Count);
+            GlobalAnalyzerConfigBuilder globalAnalyzerConfigBuilder = new GlobalAnalyzerConfigBuilder();
+            foreach (var config in analyzerConfigs)
+            {
+                if (IsGlobalConfig(config))
+                {
+                    globalAnalyzerConfigBuilder.MergeIntoGlobalConfig(config);
+                }
+                else
+                {
+                    filteredConfigs.Add(config);
+                }
+            }
+
+            var globalConfig = globalAnalyzerConfigBuilder.Build(out unsetAnalyzerKeys);
+            filteredConfigs.Insert(0, globalConfig);
+            return filteredConfigs.ToImmutableAndFree();
+
+            bool containsGlobalConfigs()
+            {
+                foreach (var config in analyzerConfigs)
+                {
+                    if (IsGlobalConfig(config))
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+        }
+
+        internal static bool IsGlobalConfig(AnalyzerConfig config) => config.GlobalSection.Properties.ContainsKey(GlobalKey);
```

Comment:
Why does this exist at all? Let's have one function to call to check if it's a global config.